File: ios/SsdidWallet/Domain/OpenId4Vp/PresentationSubmission.swift
import Foundation

/// Describes how VP tokens map to input descriptors in a presentation definition.
struct PresentationSubmission: Codable {
    let id: String
    let definitionId: String
    let descriptorMap: [DescriptorMapEntry]

    enum CodingKeys: String, CodingKey {
        case id
        case definitionId = "definition_id"
        case descriptorMap = "descriptor_map"
    }

    /// Serializes this submission to a JSON string.
    func toJson() throws -> String {
        let encoder = JSONEncoder()
        let data = try encoder.encode(self)
        guard let json = String(data: data, encoding: .utf8) else {
            throw OpenId4VpError.invalidRequest("Failed to encode PresentationSubmission")
        }
        return json
    }

    /// Creates a PresentationSubmission for SD-JWT VCs.
    static func create(
        definitionId: String,
        descriptorIds: [String]
    ) -> PresentationSubmission {
        PresentationSubmission(
            id: UUID().uuidString,
            definitionId: definitionId,
            descriptorMap: descriptorIds.map { descId in
                DescriptorMapEntry(id: descId, format: "vc+sd-jwt", path: "$")
            }
        )
    }
}

/// A single entry in the descriptor_map of a PresentationSubmission.
struct DescriptorMapEntry: Codable {
    let id: String
    let format: String
    let path: String
}

File: android/sdk/ssdid-core/src/main/java/my/ssdid/sdk/domain/oid4vp/PresentationSubmission.kt
package my.ssdid.sdk.domain.oid4vp

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.UUID

@Serializable
data class PresentationSubmission(
    val id: String,
    @SerialName("definition_id") val definitionId: String,
    @SerialName("descriptor_map") val descriptorMap: List<DescriptorMapEntry>
) {
    fun toJson(): String = json.encodeToString(this)

    companion object {
        private val json = Json { encodeDefaults = true }

        fun create(definitionId: String, descriptorIds: List<String>): PresentationSubmission {
            return PresentationSubmission(
                id = UUID.randomUUID().toString(),
                definitionId = definitionId,
                descriptorMap = descriptorIds.map { descId ->
                    DescriptorMapEntry(id = descId, format = "vc+sd-jwt", path = "$")
                }
            )
        }
    }
}

@Serializable
data class DescriptorMapEntry(
    val id: String,
    val format: String,
    val path: String
)
